### PR TITLE
Add implementation of new DataReader APIs [10180]

### PR DESCRIPTION
**Pull request by MiguelCompany**
This PR implements the new read / take APIs of the data reader.

It also fixes some of the tests added on #1649 

I will add a follow-up PR with some examples

### CODE REVIEW

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/3436/)
* Linux aarch64[Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux%20aarch64/630/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/3062/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/3470/)

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/3445/)
* Linux aarch64[Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux%20aarch64/638/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/3066/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/3479/)

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/3447/)
* Linux aarch64[Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux%20aarch64/640/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/3068/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/3481/)

File: src/cpp/fastdds/subscriber/DataReaderImpl/SampleLoanManager.hpp [hunk 1]
```diff
@@ -0,0 +1,264 @@
...
+        id.sequence_number(change->sequenceNumber);
+
+        auto comp = [id](const OutstandingLoanItem& item)
+                {
+                    return id == item.identity;
+                };
+        auto it = std::find_if(used_loans_.begin(), used_loans_.end(), comp);
+        if (it != used_loans_.end())
+        {
+            return &(*it);
```
Comment (rsanchez15) [inline src/cpp/fastdds/subscriber/DataReaderImpl/SampleLoanManager.hpp:237]: None of the tests reach this point.

File: src/cpp/fastdds/subscriber/DataReaderImpl/SampleLoanManager.hpp [hunk 2]
```diff
@@ -0,0 +1,264 @@
...
+        auto comp = [sample](const OutstandingLoanItem& item)
+                {
+                    return sample == item.sample;
+                };
+        auto it = std::find_if(used_loans_.begin(), used_loans_.end(), comp);
+        if (it != used_loans_.end())
+        {
+            return &(*it);
+        }
+        return nullptr;
```
Comment (rsanchez15) [inline src/cpp/fastdds/subscriber/DataReaderImpl/SampleLoanManager.hpp:254]: Never reached in tests.

File: src/cpp/fastdds/subscriber/DataReaderImpl.cpp [hunk 3]
```diff
@@ -218,12 +225,412 @@ DataReaderImpl::~DataReaderImpl()
...
+            // When max_samples is LENGTH_UNLIMITED, the collection imposes the maximum number of samples
+            max_samples = collection_max;
+        }
+        else
+        {
+            if (max_samples > collection_max)
+            {
+                return ReturnCode_t::RETCODE_PRECONDITION_NOT_MET;
+            }
+        }
```
Comment (rsanchez15) [inline src/cpp/fastdds/subscriber/DataReaderImpl.cpp:283]: Missing tests

File: src/cpp/fastdds/subscriber/DataReaderImpl.cpp [hunk 4]
```diff
@@ -218,12 +225,412 @@ DataReaderImpl::~DataReaderImpl()
...
+    {
+        // Loan not required, input collections should not be already loaned
+        if (false == data_values.has_ownership())
+        {
+            return ReturnCode_t::RETCODE_PRECONDITION_NOT_MET;
+        }
+
+        uint32_t collection_input_max = data_values.maximum();
+        int32_t collection_max = (collection_input_max > std::numeric_limits<uint32_t>::max() / 2u) ?
+                std::numeric_limits<int32_t>::max() : static_cast<int32_t>(collection_input_max);
```
Comment (rsanchez15) [inline src/cpp/fastdds/subscriber/DataReaderImpl.cpp:269]: Missing tests for the second clause

File: src/cpp/fastdds/subscriber/DataReaderImpl.cpp [hunk 5]
```diff
@@ -218,12 +225,412 @@ DataReaderImpl::~DataReaderImpl()
...
+{
+    if (reader_ == nullptr)
+    {
+        return ReturnCode_t::RETCODE_NOT_ENABLED;
+    }
+
+    ReturnCode_t code = check_collection_preconditions_and_calc_max_samples(data_values, sample_infos, max_samples);
+    if (!code)
+    {
+        return code;
```
Comment (rsanchez15) [inline src/cpp/fastdds/subscriber/DataReaderImpl.cpp:388]: Missing negative test

File: src/cpp/fastdds/subscriber/DataReaderImpl.cpp [hunk 6]
```diff
@@ -218,12 +225,412 @@ DataReaderImpl::~DataReaderImpl()
...
```
Comment (rsanchez15) [inline src/cpp/fastdds/subscriber/DataReaderImpl.cpp:433]: Missing negative test

File: src/cpp/fastdds/subscriber/DataReaderImpl.cpp [hunk 7]
```diff
@@ -218,12 +225,412 @@ DataReaderImpl::~DataReaderImpl()
...
```
Comment (rsanchez15) [inline src/cpp/fastdds/subscriber/DataReaderImpl.cpp:473]: Missing negative test

File: src/cpp/fastdds/subscriber/DataReaderImpl.cpp [hunk 8]
```diff
@@ -218,12 +225,412 @@ DataReaderImpl::~DataReaderImpl()
...
```
Comment (rsanchez15) [inline src/cpp/fastdds/subscriber/DataReaderImpl.cpp:556]: Missing negative test

**Review by rsanchez15 — CHANGES_REQUESTED** [review]
Partial review: missing tests.

File: src/cpp/fastrtps_deprecated/subscriber/SubscriberHistory.cpp [hunk 9]
```diff
@@ -572,5 +572,53 @@ bool SubscriberHistory::get_next_deadline(
...
+        {
+            if (exact)
+            {
+                return { false, {InstanceHandle_t(), nullptr} };
+            }
+            InstanceHandle_t tmp;
+            tmp.value[0] = 1;
+            return { true, {tmp, &m_changes} };
+        }
+    }
```
[the remark this answers is not in the document]
Comment (MiguelCompany) [reply src/cpp/fastrtps_deprecated/subscriber/SubscriberHistory.cpp:599]: You are right, the only data type is a keyed one. I should add some tests for a keyed topic.

File: src/cpp/fastdds/subscriber/DataReaderImpl/ReadTakeCommand.hpp [hunk 10]
```diff
@@ -0,0 +1,317 @@
...
+    LoanableCollection::size_type current_slot_ = 0;
+
+    bool go_to_first_valid_instance()
+    {
+        while (!is_current_instance_valid())
+        {
+            if (!next_instance())
+            {
+                return false;
+            }
```
[the remark this answers is not in the document]
Comment (MiguelCompany) [reply src/cpp/fastdds/subscriber/DataReaderImpl/ReadTakeCommand.hpp:184]: Understood. We need to check with more than one valid instance.

File: src/cpp/fastdds/subscriber/DataReaderImpl/ReadTakeCommand.hpp [hunk 11]
```diff
@@ -0,0 +1,317 @@
...
+        auto result = history_.lookup_instance(handle_, false);
+        if (!result.first)
+        {
+            finished_ = true;
+            return false;
+        }
+
+        instance_ = result.second;
+        handle_ = instance_.first;
+        return true;
```
[the remark this answers is not in the document]
Comment (MiguelCompany) [reply src/cpp/fastdds/subscriber/DataReaderImpl/ReadTakeCommand.hpp:214]: Understood. We need to check with more than one valid instance.

File: src/cpp/fastdds/subscriber/DataReaderImpl/DataReaderLoanManager.hpp [hunk 12]
```diff
@@ -0,0 +1,152 @@
...
+        {
+            OutstandingLoanItem tmp;
+            result = used_loans_.push_back(tmp);
+            if (nullptr == result)
+            {
+                return ReturnCode_t::RETCODE_OUT_OF_RESOURCES;
+            }
+
+            result->data_values = new LoanableCollection::element_type[max_samples_];
+            result->sample_infos = new LoanableCollection::element_type[max_samples_];
```
[the remark this answers is not in the document]
Comment (MiguelCompany) [reply src/cpp/fastdds/subscriber/DataReaderImpl/DataReaderLoanManager.hpp:91]: Should be solved by ad2eff5

File: src/cpp/fastdds/subscriber/DataReaderImpl/ReadTakeCommand.hpp [hunk 13]
```diff
@@ -0,0 +1,317 @@
...
+            case eprosima::fastrtps::rtps::NOT_ALIVE_DISPOSED:
+            case eprosima::fastrtps::rtps::NOT_ALIVE_DISPOSED_UNREGISTERED:
+                info.instance_state = NOT_ALIVE_DISPOSED_INSTANCE_STATE;
+                info.valid_data = false;
+                break;
+            default:
+                //TODO [ILG] change this if the other kinds ever get implemented
+                info.instance_state = ALIVE_INSTANCE_STATE;
+                break;
+        }
```
[the remark this answers is not in the document]
Comment (MiguelCompany) [reply src/cpp/fastdds/subscriber/DataReaderImpl/ReadTakeCommand.hpp:307]: I should add a test for reading disposed instances. I was leaving it for later, when we also add full support for instances (generation, view_state, instance_state, etc.) but I think I can add a simple tests to read a disposed instance.

File: src/cpp/fastdds/subscriber/DataReaderImpl/SampleInfoPool.hpp [hunk 14]
```diff
@@ -0,0 +1,113 @@
...
+        return used_items_.size();
+    }
+
+    SampleInfo* get_item()
+    {
+        SampleInfo** result = nullptr;
+
+        if (free_items_.empty())
+        {
+            result = used_items_.push_back(new SampleInfo());
```
[the remark this answers is not in the document]
Comment (MiguelCompany) [reply src/cpp/fastdds/subscriber/DataReaderImpl/SampleInfoPool.hpp:73]: Should be checked on ad2eff5

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/3469/)
* Linux aarch64[Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux%20aarch64/661/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/3087/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/3502/)

File: src/cpp/fastdds/subscriber/DataReaderImpl/SampleInfoPool.hpp [hunk 15]
```diff
@@ -0,0 +1,113 @@
...
+
+        return result ? *result : nullptr;
+    }
+
+    ReturnCode_t return_item(
+            SampleInfo* item)
+    {
+        if (!used_items_.remove(item))
+        {
+            return ReturnCode_t::RETCODE_PRECONDITION_NOT_MET;
```
[the remark this answers is not in the document]
Comment (MiguelCompany) [reply src/cpp/fastdds/subscriber/DataReaderImpl/SampleInfoPool.hpp:91]: Looking at this again, I don't think it is possible from the user API to make this happen, so I will turn this into an assertion.

File: src/cpp/fastdds/subscriber/DataReaderImpl/SampleLoanManager.hpp [hunk 16]
```diff
@@ -0,0 +1,264 @@
...
+            item = used_loans_.push_back({});
+            if (nullptr != item)
+            {
+                // Create sample if necessary
+                if (!type_->is_plain())
+                {
+                    item->sample = type_->createData();
+                }
+            }
+        }
```
[the remark this answers is not in the document]
Comment (MiguelCompany) [reply src/cpp/fastdds/subscriber/DataReaderImpl/SampleLoanManager.hpp:108]: Should be checked on 4aaac98

File: src/cpp/fastdds/subscriber/DataReaderImpl/SampleLoanManager.hpp [hunk 17]
```diff
@@ -0,0 +1,264 @@
...
+            CacheChange_t* change,
+            void*& sample)
+    {
+        OutstandingLoanItem* item = find_by_change(change);
+        if (nullptr != item)
+        {
+            item->num_refs += 1;
+            sample = item->sample;
+            return ReturnCode_t::RETCODE_OK;
+        }
```
[the remark this answers is not in the document]
Comment (MiguelCompany) [reply src/cpp/fastdds/subscriber/DataReaderImpl/SampleLoanManager.hpp:94]: I think you mean there is no test for `item != nullptr`, as on the first loan `find_by_change` will return `nullptr`. PR #1673 adds a test that will loan the same sample more than once, so this part of the code will be covered.

File: src/cpp/fastdds/subscriber/DataReaderImpl/SampleLoanManager.hpp [hunk 18]
```diff
@@ -0,0 +1,264 @@
...
+
+    ReturnCode_t return_loan(
+            void* sample)
+    {
+        OutstandingLoanItem* item = find_by_sample(sample);
+
+        if (nullptr == item)
+        {
+            return ReturnCode_t::RETCODE_BAD_PARAMETER;
+        }
```
[the remark this answers is not in the document]
Comment (MiguelCompany) [reply src/cpp/fastdds/subscriber/DataReaderImpl/SampleLoanManager.hpp:163]: Looking at this again, I don't think it is possible from the user API to make this happen, so I will turn this into an assertion.

File: src/cpp/fastdds/subscriber/DataReaderImpl/SampleLoanManager.hpp (hunk 1 above)
Comment (MiguelCompany) [reply src/cpp/fastdds/subscriber/DataReaderImpl/SampleLoanManager.hpp:237]: PR #1673 adds a test that will loan the same sample more than once, so this part of the code will be covered.

File: src/cpp/fastdds/subscriber/DataReaderImpl/SampleLoanManager.hpp (hunk 2 above)
Comment (MiguelCompany) [reply src/cpp/fastdds/subscriber/DataReaderImpl/SampleLoanManager.hpp:254]: I don't think it is possible from the user API to make this happen, so I will turn the condition on line 250 into an assertion.

File: src/cpp/fastdds/subscriber/DataReaderImpl.cpp (hunk 4 above)
Comment (MiguelCompany) [reply src/cpp/fastdds/subscriber/DataReaderImpl.cpp:269]: I would say the missing test is for the first one (when the input collection has a maximum value above the maximum of a signed integer).

File: src/cpp/fastdds/subscriber/DataReaderImpl.cpp (hunk 5 above)
Comment (MiguelCompany) [reply src/cpp/fastdds/subscriber/DataReaderImpl.cpp:388]: This should be checked by `collection_preconditions` test.

File: src/cpp/fastdds/subscriber/DataReaderImpl.cpp (hunk 6 above)
Comment (MiguelCompany) [reply src/cpp/fastdds/subscriber/DataReaderImpl.cpp:433]: This should be checked by `collection_preconditions` test.

File: src/cpp/fastdds/subscriber/DataReaderImpl.cpp (hunk 7 above)
Comment (MiguelCompany) [reply src/cpp/fastdds/subscriber/DataReaderImpl.cpp:473]: This should be checked by `collection_preconditions` test.

File: src/cpp/fastdds/subscriber/DataReaderImpl.cpp (hunk 8 above)
Comment (MiguelCompany) [reply src/cpp/fastdds/subscriber/DataReaderImpl.cpp:556]: This should be checked by `collection_preconditions` test.

File: src/cpp/fastdds/subscriber/DataReaderImpl/SampleInfoPool.hpp (hunk 15 above)
[the remark this answers is not in the document]
Comment (MiguelCompany) [reply src/cpp/fastdds/subscriber/DataReaderImpl/SampleInfoPool.hpp]: Done in 6addbc6

File: src/cpp/fastdds/subscriber/DataReaderImpl/SampleLoanManager.hpp (hunk 18 above)
[the remark this answers is not in the document]
Comment (MiguelCompany) [reply src/cpp/fastdds/subscriber/DataReaderImpl/SampleLoanManager.hpp]: Done in 6addbc6

File: src/cpp/fastdds/subscriber/DataReaderImpl/SampleLoanManager.hpp (hunk 2 above)
Comment (MiguelCompany) [reply src/cpp/fastdds/subscriber/DataReaderImpl/SampleLoanManager.hpp]: Done in 6addbc6

File: src/cpp/fastdds/subscriber/DataReaderImpl.cpp (hunk 3 above)
Comment (MiguelCompany) [reply src/cpp/fastdds/subscriber/DataReaderImpl.cpp:283]: Added on 9acab07

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/3470/)
* Linux aarch64[Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux%20aarch64/662/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/3088/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/3503/)

File: src/cpp/fastdds/subscriber/DataReaderImpl/SampleLoanManager.hpp [hunk 19]
```diff
@@ -0,0 +1,264 @@
...
+            // Reuse a free entry
+            item = used_loans_.push_back(free_loans_.back());
+            assert(nullptr != item);
+            free_loans_.pop_back();
+        }
+
+        // Early return if could not find an entry
+        if (nullptr == item)
+        {
+            return ReturnCode_t::RETCODE_OUT_OF_RESOURCES;
```
[the remark this answers is not in the document]
Comment (MiguelCompany) [reply src/cpp/fastdds/subscriber/DataReaderImpl/SampleLoanManager.hpp]: I refactored this on af9e0be and a766f5c

File: src/cpp/fastdds/subscriber/DataReaderImpl.cpp [hunk 20]
```diff
@@ -218,12 +225,412 @@ DataReaderImpl::~DataReaderImpl()
...
+        return ReturnCode_t::RETCODE_NOT_ENABLED;
+    }
+
+    // Properties should be the same on both collections
+    if ((data_values.has_ownership() != sample_infos.has_ownership()) ||
+            (data_values.maximum() != sample_infos.maximum()) ||
+            (data_values.length() != sample_infos.length()))
+    {
+        return ReturnCode_t::RETCODE_PRECONDITION_NOT_MET;
+    }
```
[the remark this answers is not in the document]
Comment (MiguelCompany) [reply src/cpp/fastdds/subscriber/DataReaderImpl.cpp]: Done in fd73965

File: src/cpp/fastdds/subscriber/DataReaderImpl.cpp [hunk 21]
```diff
@@ -218,12 +225,412 @@ DataReaderImpl::~DataReaderImpl()
...
+    {
+        return code;
+    }
+
+    auto instance = history_.lookup_instance(HANDLE_NIL, false).second;
+    detail::StateFilter states{ sample_states, view_states, instance_states };
+    detail::ReadTakeCommand cmd(*this, data_values, sample_infos, max_samples, states, instance);
+    while (!cmd.is_finished())
+    {
+        cmd.add_instance(false);
```
[the remark this answers is not in the document]
Comment (MiguelCompany) [reply src/cpp/fastdds/subscriber/DataReaderImpl.cpp]: Done in 324115c

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/3479/)
* Linux aarch64[Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux%20aarch64/668/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/3095/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/3510/)

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/3481/)
* Linux aarch64[Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux%20aarch64/670/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/3097/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/3512/)

File: src/cpp/fastdds/subscriber/DataReaderImpl/SampleLoanManager.hpp (hunk 16 above)
[the remark this answers is not in the document]
Comment (rsanchez15) [reply src/cpp/fastdds/subscriber/DataReaderImpl/SampleLoanManager.hpp:116]: These lines are still not covered.

File: src/cpp/fastdds/subscriber/DataReaderImpl/SampleLoanManager.hpp [hunk 22]
```diff
@@ -0,0 +1,258 @@
...
+        // Perform deserialization
+        if (type_->is_plain())
+        {
+            auto ptr = item->payload.data;
+            ptr += item->payload.representation_header_size;
+            item->sample = ptr;
+        }
+        else
+        {
+            type_->deserialize(&item->payload, item->sample);
```
Comment (rsanchez15) [inline src/cpp/fastdds/subscriber/DataReaderImpl/SampleLoanManager.hpp:150]: Not tested

**Review by rsanchez15 — CHANGES_REQUESTED** [review]

File: src/cpp/fastdds/subscriber/DataReaderImpl/SampleLoanManager.hpp [hunk 23]
```diff
@@ -0,0 +1,258 @@
...
+            for (const OutstandingLoanItem& item : free_loans_)
+            {
+                type_->deleteData(item.sample);
+            }
+        }
+    }
+
+    int32_t num_allocated() const
+    {
+        assert(used_loans_.size() <= std::numeric_limits<int32_t>::max());
```
[the remark this answers is not in the document]
Comment (MiguelCompany) [reply src/cpp/fastdds/subscriber/DataReaderImpl/SampleLoanManager.hpp:86]: It bothers me that the warning didn't show up on the CI. Done on a9e221a.

File: src/cpp/fastrtps_deprecated/subscriber/SubscriberHistory.cpp (hunk 9 above)
[the remark this answers is not in the document]
Comment (MiguelCompany) [reply src/cpp/fastrtps_deprecated/subscriber/SubscriberHistory.cpp:599]: I will add tests for non-keyed types on #1673

File: src/cpp/fastdds/subscriber/DataReaderImpl/ReadTakeCommand.hpp [hunk 24]
```diff
@@ -0,0 +1,317 @@
...
+    }
+
+    bool add_instance(
+            bool take_samples)
+    {
+        // Advance to the first instance with a valid state
+        if (!go_to_first_valid_instance())
+        {
+            return false;
```
[the remark this answers is not in the document]
Comment (MiguelCompany) [reply src/cpp/fastdds/subscriber/DataReaderImpl/ReadTakeCommand.hpp:96]: We will add this on another PR.

File: src/cpp/fastdds/subscriber/DataReaderImpl/ReadTakeCommand.hpp (hunk 10 above)
[the remark this answers is not in the document]
Comment (MiguelCompany) [reply src/cpp/fastdds/subscriber/DataReaderImpl/ReadTakeCommand.hpp:184]: We will add this on another PR.

File: src/cpp/fastdds/subscriber/DataReaderImpl/ReadTakeCommand.hpp (hunk 11 above)
[the remark this answers is not in the document]
Comment (MiguelCompany) [reply src/cpp/fastdds/subscriber/DataReaderImpl/ReadTakeCommand.hpp:214]: I will add tests for two valid instances on #1673

File: src/cpp/fastdds/subscriber/DataReaderImpl/ReadTakeCommand.hpp (hunk 13 above)
[the remark this answers is not in the document]
Comment (MiguelCompany) [reply src/cpp/fastdds/subscriber/DataReaderImpl/ReadTakeCommand.hpp:307]: I will add tests for reading disposed instances on #1673

File: src/cpp/fastdds/subscriber/DataReaderImpl/SampleLoanManager.hpp (hunk 16 above)
[the remark this answers is not in the document]
Comment (MiguelCompany) [reply src/cpp/fastdds/subscriber/DataReaderImpl/SampleLoanManager.hpp:116]: Should be with f81a079

File: src/cpp/fastdds/subscriber/DataReaderImpl/SampleLoanManager.hpp [hunk 25]
```diff
@@ -0,0 +1,264 @@
...
+            if (type_->is_plain())
+            {
+                auto ptr = item->payload.data;
+                ptr += item->payload.representation_header_size;
+                item->sample = ptr;
+            }
+            else
+            {
+                type_->deserialize(&item->payload, item->sample);
+            }
```
[the remark this answers is not in the document]
Comment (MiguelCompany) [reply src/cpp/fastdds/subscriber/DataReaderImpl/SampleLoanManager.hpp]: I will add tests for non-plain types on #1673

File: src/cpp/fastdds/subscriber/DataReaderImpl.cpp (hunk 4 above)
Comment (MiguelCompany) [reply src/cpp/fastdds/subscriber/DataReaderImpl.cpp:276]: Adding a test for this would mean using a collection with space for 2^31 items. It may use up 520 GB of memory.

File: src/cpp/fastdds/subscriber/DataReaderImpl/SampleLoanManager.hpp (hunk 22 above)
Comment (MiguelCompany) [reply src/cpp/fastdds/subscriber/DataReaderImpl/SampleLoanManager.hpp:150]: I will add tests for non-plain types on #1673

File: src/cpp/fastdds/subscriber/DataReaderImpl/ReadTakeCommand.hpp [hunk 26]
```diff
@@ -0,0 +1,317 @@
...
+        finished_ = nullptr == instance.second;
+    }
+
+    ~ReadTakeCommand()
+    {
+        if (!data_values_.has_ownership() && ReturnCode_t::RETCODE_NO_DATA == return_value_)
+        {
+            loan_manager_.return_loan(data_values_, sample_infos_);
+            data_values_.unloan();
+            sample_infos_.unloan();
```
[the remark this answers is not in the document]
Comment (MiguelCompany) [reply src/cpp/fastdds/subscriber/DataReaderImpl/ReadTakeCommand.hpp:86]: Will be covered by a new test on #1673

File: src/cpp/fastdds/subscriber/DataReaderImpl.cpp (hunk 4 above)
Comment (MiguelCompany) [reply src/cpp/fastdds/subscriber/DataReaderImpl.cpp:276]: I have just thought of a way to check this without using a huge amount of memory. Will implement it after lunch.

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/3483/)
* Linux aarch64[Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux%20aarch64/672/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/3099/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/3514/)

**Review by rsanchez15 — APPROVED** [review]
